### DISCUSSION TITLE
Any faster way to join images?

### DISCUSSION THREAD

**Discussion by 60215** [libvips/libvips#3247] [Help wanted]
I wrote a program with pillow to make an image out of other images. I wanted to rewrite it on c++ for performance reasons, as well as c++ being better suited for later parts of the program. It joins 6k+ images. However, I noticed that pillow's .paste() on canvas works faster than libvips_arrayjoin(merge and insert are even slower than arrayjoin) and was wondering if there is faster way to do this. Here's my code in c++:

```
#include <vips/vips8>
#include <iostream>

int main( int argc, char* args[] ) {
    std::cout << VIPS_INIT(args[0]) << std::endl;

    int WIDTH = 1920;
    int HEIGHT = 1080;
    int parts = 120;
    int n = 5;
    float factorX;
    float factorY;

    VipsImage* out;

    VipsImage* arr[5];
    arr[0] = vips_image_new_from_file("1.png", "access", VIPS_ACCESS_SEQUENTIAL, NULL);
    arr[1] = vips_image_new_from_file("2.png", "access", VIPS_ACCESS_SEQUENTIAL, NULL);
    arr[2] = vips_image_new_from_file("3.png", "access", VIPS_ACCESS_SEQUENTIAL, NULL);
    arr[3] = vips_image_new_from_file("4.png", "access", VIPS_ACCESS_SEQUENTIAL, NULL);
    arr[4] = vips_image_new_from_file("5.png", "access", VIPS_ACCESS_SEQUENTIAL, NULL);

    for(int i = 0; i < 5; i++) {
        factorX = (WIDTH/(float)parts)/vips_image_get_width(arr[i]);
        factorY = (HEIGHT/(float)parts)/vips_image_get_height(arr[i]);
        vips_resize(arr[i], &arr[i], factorX, "vscale", factorY, NULL);
    }

    VipsImage* tojoin[parts*parts];

    //generate image 5 times
    for(int k = 0; k < 5; k++) {
        //generate array for image
        for (int i = 0; i < parts; i++) {
            for (int j = 0; j < parts; j++) {
                tojoin[i * parts + j] = arr[(i+j) % 5];
            }
        }
        vips_arrayjoin(tojoin, &out, parts*parts, "across", parts, NULL);
    }

    vips_image_write_to_file(out, "final.png", NULL);
}
```
and here's my code in python:
```
from PIL import Image

WIDTH = 1920
HEIGHT = 1080
parts = 120

small_x = int (WIDTH/parts)
small_y = int (HEIGHT/parts)

bg = Image.new("RGB", (1920, 1080))


images = [Image.open("1.png").resize((small_x, small_y)),
          Image.open("2.png").resize((small_x, small_y)),
          Image.open("3.png").resize((small_x, small_y)),
          Image.open("4.png").resize((small_x, small_y)),
          Image.open("5.png").resize((small_x, small_y)),
         ]

#generate image 5 times
for k in range(5):
    #generate image
    for i in range(parts):
        for j in range(parts):
            bg.paste(images[(i + j)%5], (small_x*i, small_y*j))

bg.save("final.png")
```
Pillow works faster in this case and I wondered if there is a way to speed libvips up. Perhaps using tiff's might help, I am quite new to the image processing.
For me python works ~0.15 second and c takes 0.8 seconds timing only the pasting/arrayjoining part.
In actual program images may be located differently, randomly.
Am I doing something wrong or is libvips not suited for this task?

**Comment by jcupitt** [comment]
Hi @60215,

Wow you're right `arrayjoin` performs really poorly in this case. I'll see if I can see what's going wrong.

**Reply by jcupitt** [reply]
... the pyvips test program I made:

```python
#!/usr/bin/python3

import pyvips

WIDTH = 1920
HEIGHT = 1080
parts = 120

small_x = int (WIDTH/parts)
small_y = int (HEIGHT/parts)

images = [pyvips.Image
            .thumbnail(f"{i}.png", small_x, height=small_y, size="force")
            .copy_memory()
          for i in range(1, 6)]

tojoin = [images[(x + y) % 5] for x in range(parts) for y in range(parts)]

final = pyvips.Image.arrayjoin(tojoin, across=parts) 
print(f"writing {final} ...")
final.write_to_file("final.png")
```

**Reply by 60215** [reply]
Thank you for the reply. This works, however, it has performance comparable to C. In original example I ran image generation five times, but without it my old python script works in 0.09 seconds, and pyvips one in 0.16 seconds(with writing to file up to 10 seconds). I guess the difference isn't that big and the bottleneck is in another part of the program. I could swear I saw a better time first time I ran it, but whatever. The nine second writing to file time is not very usable though.
I will try to look at alternatives around.

**Reply by jcupitt** [reply]
That was just the test prog I made to reproduce your issue, not a fix. I'm looking into the cause now ... it looks like you've hit a strange O(N^2) case in `arrayjoin`. I'll update if I find it.

**Reply by jcupitt** [reply]
I improved it a bit with https://github.com/libvips/libvips/commit/feae09e9cdc06a6667bc209d655f9c4bc78ddb4a

I now see:

```
$ time ../join-pil.py 

real	0m0.363s
user	0m0.355s
sys	0m0.008s

$ time ../join-vips.py 
writing <pyvips.Image 3072x3072 uchar, 3 bands, srgb> to final.png ...

real	0m0.689s
user	0m0.670s
sys	0m0.128s
```

With this test prog:

```python
#!/usr/bin/python3

import pyvips

WIDTH = 1920
HEIGHT = 1080
parts = 120

small_x = int (WIDTH/parts)
small_y = int (HEIGHT/parts)

images = [pyvips.Image
            .thumbnail(f"{i}.png", small_x, height=small_y, size="force")
            .copy_memory()
          for i in range(1, 6)]
        
tojoin = [images[i % 5] for i in range(parts ** 2)]

final = pyvips.Image.arrayjoin(tojoin, across=parts) 
print(f"writing {final} to final.png ...")
final.write_to_file("final.png")
```

So still 2x slower, but it's at least closer now. I'll see if I can find anything else to fix.

If I change them to make a 19,200 x 10,800 pixel image, I see:

```
$ time ../join-pil.py 

real	0m6.425s
user	0m6.172s
sys	0m0.252s
$ time ../join-vips.py 
writing <pyvips.Image 19200x10800 uchar, 3 bands, srgb> to final.png ...

real	0m2.982s
user	0m4.990s
sys	0m2.008s
```

And the libvips one uses 150mb of memory, vs. almost 1gb for PIL.

**Reply by jcupitt** [reply]
I merged this to 8.14 as well, so this improvement will be in the new release. Thanks for pointing out this problem!

**Reply by 60215** [reply]
Thank you for improving this!
